### ISSUE TITLE
Methods to help coding against Roslyn.

### ISSUE THREAD

**Issue by AdamSpeight2008** [DotNetAnalyzers/Proposals#31]
Whilst creating these analysers, we find a common actions that can be extracted out into a separate package(s). That further help use to code roslyn based diagnostics.

For example

``` vb
Namespace Global.RoslynExts
   Namespace VB
    <HideModuleName()>
    Public Module Exts

      <Extension>
      Public Function InWhichMethod(sn As SyntaxNode) As MethodBaseSyntax
        Return sn.AncestorsAndSelf.OfType(Of MethodBaseSyntax).FirstOrDefault
      End Function

      <Extension>
      Public Function InWhichTypeDeclaration(sn As SyntaxNode) As TypeBlockSyntax
        Return sn.AncestorsAndSelf.OfType(of TypeBlockSyntax).FirstOrDefault
      End Function
    End Module

  End Namespace

End Namespace
```

**Comment by AdamSpeight2008** [comment]
Pseudo-Template Functions

``` vb.net
Public Module Exts
  <Runtime.CompilerServices.Extension>
  Public Function ToExSt(of T As StatementSyntax)(code As string) As T
    Return DirectCast(ParseExecutableStatement(code),T)
  End Function

  <Runtime.CompilerServices.Extension>
  Public Function ToEx(Of T As ExpressionSyntax)(code As String) As T
    Return DirectCast(ParseExpression(code), T)
  End Function

End Module
```

**Comment by AdamSpeight2008** [comment]
Implementation started in https://github.com/AdamSpeight2008/RoslynExts

**Comment by sharwell** [comment]
**tl;dr** I would like a bit more time to review this one, but if you don't want to wait then you can always convince the other owners of the organization to bring the code in now.

I agree that currently it can be challenging to work with the .NET Compiler Platform. However, I'm not sure what the best approach is for resolving the situation. Since it's early on, many of the problem areas might be best solved by reporting issues/suggestions to the [.NET Compiler Platform issue tracker](https://roslyn.codeplex.com/workitem/list/basic) so they might be included in a future release.

In other cases, it's not clear what kinds of best practices will lead to the best user experience (especially regarding performance and correctness).

I'd like to wait up to a couple months while we work on a variety of actual analyzer implementations to see what patterns emerge before refactoring the truly shared functionality into a general library.

**Comment by AdamSpeight2008** [comment]
"If you build it they will come"
I'm starting one so I can create a nuget package that the others can reference. Along the way to finding out the best practices, it can be useful to have help on the way. Other contributors can add extensions and method they find useful (and others could to).
